### PR TITLE
Add lettuce 5.2 instrumentation that uses lettuce's native tracing functionality.

### LINKED ISSUE #408: Capture redis key accessed in lettuce autoinstrumentation
When lettuce is used to access redis, the trace just shows the type of operation but it doesn't show specifics.

I am wondering if it would be possible to include the relevant redis key as an attribute.

### LINKED ISSUE #504: lettuce-5.0 tests run only on java 8
If you look at a build, for example here

https://app.circleci.com/pipelines/github/open-telemetry/opentelemetry-java-instrumentation/1291/workflows/6aa5dcc9-cad7-49fa-8415-f3757838a2fe/jobs/10525/steps

you'll notice `lettuce-5.0:test` is `SKIPPED`, unlike others

```
> Task :instrumentation:jedis:jedis-1.4:test
> Task :instrumentation:lettuce:lettuce-5.0:compileTestGroovy UP-TO-DATE
> Task :instrumentation:lettuce:lettuce-5.0:processTestResources NO-SOURCE
> Task :instrumentation:lettuce:lettuce-5.0:testClasses UP-TO-DATE
> Task :instrumentation:lettuce:lettuce-5.0:test SKIPPED
> Task :instrumentation:log4j:log4j-1.1:compileTestJava NO-SOURCE
> Task :instrumentation:log4j:log4j-1.1:compileTestGroovy UP-TO-DATE
> Task :instrumentation:log4j:log4j-1.1:processTestResources
> Task :instrumentation:log4j:log4j-1.1:testClasses
> Task :instrumentation:jedis:jedis-3.0:test
```

Locally, I see this

```
> Task :instrumentation:lettuce:lettuce-5.0:test SKIPPED
Skipping task ':instrumentation:lettuce:lettuce-5.0:test' as task onlyIf is false.
:instrumentation:lettuce:lettuce-5.0:test (Thread[Execution worker for ':' Thread 6,5,main]) completed. Took 0.0 secs.
```

Digged the Gradle configs some but couldn't find why this would happen.

As a trackback, also asked on Gitter about this issue but filing to make sure it's not lost - https://gitter.im/open-telemetry/opentelemetry-java-instrumentation?at=5ee34e3e7c64f31f11577744

### PR DESCRIPTION
**Pull request by anuraaga**
If muzzle lets me through in one shot I'd be amazed :)

While `Tracing` was added in 5.1, including the command was added in 5.2, and it doesn't seeem worth branching on this since this is new instrumentation, I just went ahead and targeted 5.2.

Differences in tracing
- connection and some system commands aren't traced.
- net / command tags are filled in. A couple of events that lettuce provides are too
- No corner cases, while I preserved the async/reactive tests, there isn't as much value since there is no special-casing for the APIs

Fixes #408 
Fixes #504

### CODE REVIEW

File: instrumentation/lettuce/lettuce-5.2/lettuce-5.2.gradle [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+testSets {
+  latestDepTest {
+    dirName = 'test'
+  }
+}
+
+dependencies {
+  compileOnly group: 'io.lettuce', name: 'lettuce-core', version: '5.2.0.RELEASE'
+  compileOnly group: 'io.lettuce', name: 'lettuce-core', version: '5.2.0.RELEASE'
```
Comment (iNikem) [inline instrumentation/lettuce/lettuce-5.2/lettuce-5.2.gradle:26]: Duplicate dep?

File: instrumentation/lettuce/lettuce-5.2/src/main/java8/io/opentelemetry/auto/instrumentation/lettuce/v5_2/LettuceClientDecorator.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+package io.opentelemetry.auto.instrumentation.lettuce.v5_2;
+
+import io.opentelemetry.OpenTelemetry;
+import io.opentelemetry.trace.Tracer;
+
+// Currently don't use decorator pattern since lifecycle of lettuce Tracing is too different than
+// what we expect.
+public class LettuceClientDecorator {
```
Comment (iNikem) [inline instrumentation/lettuce/lettuce-5.2/src/main/java8/io/opentelemetry/auto/instrumentation/lettuce/v5_2/LettuceClientDecorator.java:24]: Do you need this class at all?

File: instrumentation/lettuce/lettuce-5.2/src/main/java8/io/opentelemetry/auto/instrumentation/lettuce/v5_2/OpenTelemetryTracing.java [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+
+    @Nullable private Status status;
+
+    @Nullable private Span span;
+
+    protected OpenTelemetrySpan(@Nullable Span parent) {
+      this.parent = parent;
+    }
+
+    // Called before start. We need to buffer because this until then.
```
Comment (iNikem) [inline instrumentation/lettuce/lettuce-5.2/src/main/java8/io/opentelemetry/auto/instrumentation/lettuce/v5_2/OpenTelemetryTracing.java:166]: "this until then"? :)

File: instrumentation/lettuce/lettuce-5.2/src/main/java8/io/opentelemetry/auto/instrumentation/lettuce/v5_2/OpenTelemetryTracing.java [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+
+    @Override
+    public synchronized Tracer.Span start() {
+      // If name() wasn't called yet we will update it later.
+      String spanName = name != null ? name : "REDIS";
+      Span.Builder builder = TRACER.spanBuilder(spanName).setSpanKind(Kind.CLIENT);
+      if (parent != null) {
+        builder.setParent(parent);
+      } else {
+        builder.setNoParent();
```
Comment (iNikem) [inline instrumentation/lettuce/lettuce-5.2/src/main/java8/io/opentelemetry/auto/instrumentation/lettuce/v5_2/OpenTelemetryTracing.java:198]: If we don't call this, then we _may_ get a span from current Otel context, don't you think?

File: instrumentation/lettuce/lettuce-5.2/src/main/java8/io/opentelemetry/auto/instrumentation/lettuce/v5_2/OpenTelemetryTracing.java [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+        tags.add(value);
+      }
+      return this;
+    }
+
+    @Override
+    public synchronized Tracer.Span error(Throwable throwable) {
+      // TODO(anuraaga): Check if any lettuce exceptions map well to a Status and try mapping.
+      final Status status =
+          Status.INTERNAL.withDescription(throwable.getClass() + ": " + throwable.getMessage());
```
Comment (iNikem) [inline instrumentation/lettuce/lettuce-5.2/src/main/java8/io/opentelemetry/auto/instrumentation/lettuce/v5_2/OpenTelemetryTracing.java:266]: Reading javadoc, it seems to be that `UNKNOWN` suits better here?

File: instrumentation/lettuce/lettuce-5.2/src/test/groovy/io/opentelemetry/auto/instrumentation/lettuce/v5_2/LettuceAsyncClientTest.groovy [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+
+  def "connect using get on ConnectionFuture"() {
+    setup:
+    RedisClient testConnectionClient = RedisClient.create(embeddedDbUri)
+    testConnectionClient.setOptions(CLIENT_OPTIONS)
+
+    when:
+    ConnectionFuture connectionFuture = testConnectionClient.connectAsync(StringCodec.UTF8,
+      new RedisURI(HOST, port, 3, TimeUnit.SECONDS))
+    StatefulConnection connection = connectionFuture.get()
```
Comment (iNikem) [inline instrumentation/lettuce/lettuce-5.2/src/test/groovy/io/opentelemetry/auto/instrumentation/lettuce/v5_2/LettuceAsyncClientTest.groovy:123]: This can block indefinitely, not good

File: instrumentation/lettuce/lettuce-5.2/src/test/groovy/io/opentelemetry/auto/instrumentation/lettuce/v5_2/LettuceAsyncClientTest.groovy [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+
+  def "connect exception inside the connection future"() {
+    setup:
+    RedisClient testConnectionClient = RedisClient.create(dbUriNonExistent)
+    testConnectionClient.setOptions(CLIENT_OPTIONS)
+
+    when:
+    ConnectionFuture connectionFuture = testConnectionClient.connectAsync(StringCodec.UTF8,
+      new RedisURI(HOST, incorrectPort, 3, TimeUnit.SECONDS))
+    StatefulConnection connection = connectionFuture.get()
```
Comment (iNikem) [inline instrumentation/lettuce/lettuce-5.2/src/test/groovy/io/opentelemetry/auto/instrumentation/lettuce/v5_2/LettuceAsyncClientTest.groovy:142]: same here

File: instrumentation/lettuce/lettuce-5.2/src/test/groovy/io/opentelemetry/auto/instrumentation/lettuce/v5_2/LettuceAsyncClientTest.groovy [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+        span(0) {
+          operationName "SET"
+          spanKind CLIENT
+          errored false
+          tags {
+            "net.transport" "IP.TCP"
+            "net.peer.ip" "127.0.0.1"
+            "net.peer.port" port
+            "db.type" "redis"
+            "db.statement" "key<TESTSETKEY> value<TESTSETVAL>"
```
Comment (iNikem) [inline instrumentation/lettuce/lettuce-5.2/src/test/groovy/io/opentelemetry/auto/instrumentation/lettuce/v5_2/LettuceAsyncClientTest.groovy:169]: Can you fill more attributes? https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md

File: instrumentation/lettuce/lettuce-5.2/src/main/java8/io/opentelemetry/auto/instrumentation/lettuce/v5_2/OpenTelemetryTracing.java [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+  // particularly safe, synchronizing all accesses. Relying on implementation details would allow
+  // reducing synchronization but the impact should be minimal.
+  static class OpenTelemetrySpan extends Tracer.Span {
+    @Nullable private final Span parent;
+
+    @Nullable private OpenTelemetryEndpoint endpoint;
+    @Nullable private String name;
+
+    @Nullable private List<Object> events;
+    @Nullable private List<String> tags;
```
Comment (trask) [inline instrumentation/lettuce/lettuce-5.2/src/main/java8/io/opentelemetry/auto/instrumentation/lettuce/v5_2/OpenTelemetryTracing.java:156]: what do u think about replacing `parent`, `endpoint`, `name` and `tags` with `Span.Builder`?

File: instrumentation/lettuce/lettuce-5.2/src/test/groovy/io/opentelemetry/auto/instrumentation/lettuce/v5_2/LettuceAsyncClientTest.groovy [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+          }
+          event(0) {
+            eventName "redis.encode.start"
+          }
+          event(1) {
+            eventName "redis.encode.end"
+          }
```
Comment (trask) [inline instrumentation/lettuce/lettuce-5.2/src/test/groovy/io/opentelemetry/auto/instrumentation/lettuce/v5_2/LettuceAsyncClientTest.groovy:176]: are these events worth capturing, given they are not part of the [semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md)?

File: instrumentation/lettuce/lettuce-5.1/lettuce-5.1.gradle [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+testSets {
+  latestDepTest {
+    dirName = 'test'
+  }
+}
+
+dependencies {
+  compileOnly group: 'io.lettuce', name: 'lettuce-core', version: '5.2.0.RELEASE'
```
Comment (iNikem) [inline instrumentation/lettuce/lettuce-5.1/lettuce-5.1.gradle:25]: Shouldn't this be `5.1.0.RELEASE`?

File: instrumentation/lettuce/lettuce-5.1/src/test/groovy/io/opentelemetry/auto/instrumentation/lettuce/v5_1/LettuceReactiveClientTest.groovy [hunk 12]
```diff
@@ -0,0 +1,473 @@
...
+        span(0) {
+          operationName "RANDOMKEY"
+          spanKind CLIENT
+          errored false
+          tags {
+            "net.transport" "IP.TCP"
+            "net.peer.ip" "127.0.0.1"
+            "net.peer.port" port
+            "db.url" "redis://127.0.0.1:$port"
+            "db.type" "redis"
```
Comment (iNikem) [inline instrumentation/lettuce/lettuce-5.1/src/test/groovy/io/opentelemetry/auto/instrumentation/lettuce/v5_1/LettuceReactiveClientTest.groovy:235]: `db.instance` and `db.statement` are marked as required in https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md

File: instrumentation/lettuce/lettuce-5.1/src/test/groovy/io/opentelemetry/auto/instrumentation/lettuce/v5_1/LettuceAsyncClientTest.groovy [hunk 13]
```diff
@@ -0,0 +1,406 @@
...
+
+  def "connect using get on ConnectionFuture"() {
+    setup:
+    RedisClient testConnectionClient = RedisClient.create(embeddedDbUri)
+    testConnectionClient.setOptions(CLIENT_OPTIONS)
+
+    when:
+    ConnectionFuture connectionFuture = testConnectionClient.connectAsync(StringCodec.UTF8,
+      new RedisURI(HOST, port, 3, TimeUnit.SECONDS))
+    StatefulConnection connection = connectionFuture.get()
```
Comment (iNikem) [inline instrumentation/lettuce/lettuce-5.1/src/test/groovy/io/opentelemetry/auto/instrumentation/lettuce/v5_1/LettuceAsyncClientTest.groovy:123]: I don't think that all our tests has inherent timeout. At least I have seen test hanging for a very long time before I abort them manually.

File: instrumentation/lettuce/lettuce-5.1/src/test/groovy/io/opentelemetry/auto/instrumentation/lettuce/v5_1/LettuceReactiveClientTest.groovy (hunk 12 above)
Comment (anuraaga) [reply instrumentation/lettuce/lettuce-5.1/src/test/groovy/io/opentelemetry/auto/instrumentation/lettuce/v5_1/LettuceReactiveClientTest.groovy:236]: Thanks, made sure statement is always filled (realized statement doc has a template for redis that includes the span name) and filed an issue to consider making instance optional since it seems like it wouldn't always apply (redis cluster has no db index concept anyways) https://github.com/open-telemetry/opentelemetry-specification/issues/658

File: instrumentation/lettuce/lettuce-5.1/src/test/groovy/io/opentelemetry/auto/instrumentation/lettuce/v5_1/LettuceAsyncClientTest.groovy (hunk 13 above)
Comment (anuraaga) [reply instrumentation/lettuce/lettuce-5.1/src/test/groovy/io/opentelemetry/auto/instrumentation/lettuce/v5_1/LettuceAsyncClientTest.groovy:123]: Hmm - unfortunately `ConnectionFuture` doesn't offer any method that accepts a timeout :( Can I leave it like this since it's the same issue as the `lettuce-5.0` instrumentation?

File: instrumentation/lettuce/lettuce-5.1/lettuce-5.1.gradle (hunk 11 above)
Comment (anuraaga) [reply instrumentation/lettuce/lettuce-5.1/lettuce-5.1.gradle]: Yeah guess so

**Comment by anuraaga** [comment]
Codenarc seems to have crossed a threshold due to the package name having an underscore. But I notice this pattern throughout the repo - can we disable this check?

```

Rule Name | Priority | Line # | Source Line / Message
-- | -- | -- | --
PackageName | 2 |   | [MSG]package=io.opentelemetry.auto.instrumentation.lettuce.v5_1
PackageName | 2 |   | [MSG]package=io.opentelemetry.auto.instrumentation.lettuce.v5_1
PackageName | 2 |   | [MSG]package=io.opentelemetry.auto.instrumentation.lettuce.v5_1
PackageName | 2 |   | [MSG]package=io.opentelemetry.auto.instrumentation.lettuce.v5_1
PackageName | 2 |   | [MSG]package=io.opentelemetry.auto.instrumentation.lettuce.v5_1
PackageName | 2 |   | [MSG]package=io.opentelemetry.auto.instrumentation.lettuce.v5_1
PackageName | 2 |   | [MSG]package=io.opentelemetry.auto.instrumentation.lettuce.v5_1
PackageName | 2 |   | [MSG]package=io.opentelemetry.auto.instrumentation.lettuce.v5_1

```

```

  | PackageName | Verifies  that the package name for a class matches a regular expression. By  default it checks that the package name consists of only lowercase  letters, separated by periods. The regex property specifies the regular expression used to validate the package name. The packageNameRequired property indicates whether a package name declaration is required for all classes.
-- | -- | --
```

**Comment by trask** [comment]
nice muzzle work! 😄 

> Codenarc seems to have crossed a threshold due to the package name having an underscore. But I notice this pattern throughout the repo - can we disable this check?

yes, please do

**Comment by anuraaga** [comment]
Looks ready now except for a test flake
